### PR TITLE
Update LMSFilePicker UI to improve layout, UX

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR improves the user experience for configuring assignments to use LMS-stored files (Blackboard or Canvas files). This work:

* Eliminates jumpy modal resizing when file contents are updated (navigating between folders)
* Improves resiliency of modal and file table sizing in different LMS and browser contexts
* Reduces visual loading thrash by adjusting how loading spinners are handled
* Fixes a layout issue that caused a white gap and unintended top margins on modals
* Fixes a subtle one-pixel “wiggle” in the file breadcrumbs when navigating into a subdirectory
* Ensures that errors are rendered in usable ways

It replaces local `Dialog` and `Table` components with `Modal` and `Table` from the `frontend-shared` package, respectively. This will bring LMS more into consistency with our design patterns and guidelines, and brings a few usability and a11y improvements.

I’ve attempted to implement this in a relatively surgical way—there is an open issue regarding the need to refactor the `LMSFilePicker` component. I had a heavy hand with commenting in this work to help guide whomever picks up that refactoring work.

Fixes https://github.com/hypothesis/lms/issues/3022
Fixes https://github.com/hypothesis/lms/issues/3046

There is still additional work to do to make the UX of the breadcrumbs a little clearer, etc. This body of work is concerned with integrating (shared) `Table` and `Modal` into this UI flow, and the surrounding affordances needed for that.

### To test

Check out and run this branch.

#### Blackboard

1. Log in to https://aunltd-test.blackboard.com/ as `blackboardteacher`
2. Create a new assignment in [this course](https://aunltd-test.blackboard.com/ultra/courses/_19_1/cl/outline) (Content -> Build Content -> Hypothesis, localhost, make devdata) (All you need to do is give it a name and click Submit)
3. Launch the unconfigured assignment and click the “Select PDF from Blackboard” button.

#### Canvas

1. Go to the [Sections Enabled course](https://hypothesis.instructure.com/courses/125/) (Log in as eng+canvasteacher@hypothes.is)
2. In the “scratch” assignments area, add a new assignment or launch and reconfigure (“Edit Assignment Settings”) an existing assignment. The assignment’s “External Tool” needs to be configured to use the “localhost” install. A bit more info on this is [available here](https://github.com/hypothesis/lms/wiki/Testing-the-LTI-App-in-Canvas)
3. In the “External Tool” configuration, click “Select PDF from Canvas”

#### Testing the Authz flow in either Canvas or Blackboard

To reset auth token such that you get re-prompted for authorization and are able to see the auth flow in action:

1. Run the following command in your `lms` project directory: `tox -qe dockercompose -- exec postgres psql -U postgres -c 'DELETE FROM oauth2_token;'`
2. For Blackboard only, subsequently visit https://aunltd-test.blackboard.com/ultra/tools , click _Application Authorization_ and revoke Hypothesis Dev. This seems a little intimidating but is harmless.

Doing this should cause auth popups to occur in both Canvas and Blackboard when you enter into the file-picking flow.

### After these changes…

General happy-path flow in Canvas:

after-canvas-files

General happy-path flow in Blackboard (the components aren't actually off-center; I didn't center my GIF-capture frame, whoops):

after-blackboard-file-picker

Some sad path grabs:

Auth prompt, BB:

after-blackboard-authz-prompt

Authz sad, BB:

after-blackboard-authz-error

_Note_: Typography and layout rhythm could be improved in these error modals, but is beyond the scope of current work. That was untouched in these changes.

Auth prompt, Canvas:

after-canvas-authz-prompt

Authz sad, Canvas:

after-canvas-authz-error

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/ContentSelector.js [hunk 1]
```diff
@@ -236,7 +235,7 @@ export default function ContentSelector({
         </div>
         <div className="u-stretch" />
       </div>
-      {dialog}
+      <div>{dialog}</div>
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/ContentSelector.js:238]: See long commit message about this change.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.js [hunk 2]
```diff
@@ -30,15 +30,17 @@ import FileList from './FileList';
...
+      <div className="u-layout-column--align-center hyp-u-padding--6">
+        <div>
+          There are no PDFs in this {documentContext}.{' '}
+          <a href={href} target="_blank" rel="noreferrer">
+            Upload some files to the {documentContext}
+          </a>{' '}
+          and try again.
+        </div>
+      </div>
+    </Frame>
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.js:43]: Styling this was one of the hardest parts of this bit of work, oddly (per positioning/layout. Long story). A follow-up piece of work is to make the `Table` component potentially accept an optional message to show when it's empty. That would have made this crazy easier.

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.js [hunk 3]
```diff
@@ -318,20 +319,33 @@ export default function LMSFilePicker({
...
+  // error. To avoid hopscotching between modals, show only
+  // a full-screen spinner during this state.
+  const isLoading = dialogState.state === 'fetching';
+  // During subsequent file/folder fetches, represent the loading state within
+  // the file list UI (do not show a full-screen spinner)
+  const isReloading = dialogState.state === 'reloading';
+  // Render the file list and breadcrumbs UI in certain states. This also
+  // requires more real estate in the <Modal>, and applies an appropriate CSS
+  // class
+  const withFileUI = ['reloading', 'fetched'].includes(dialogState.state);
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.js:333]: Continuing to tack on state-related logic to this component is a bit inelegant, of course, but we know it's highly urgently due for a refactor, so I've left what I hope are thorough comments as to what is going on here. One of the underlying challenges is that we're asking a single component to be a whole lot of different things, and the Modal rendered below has to a lot of different things in a number of different states...

File: lms/static/scripts/frontend_apps/components/LMSFilePicker.js [hunk 4]
```diff
@@ -318,20 +319,33 @@ export default function LMSFilePicker({
...
+
+  if (isLoading) {
+    return <FullScreenSpinner />;
   }
 
   return (
-    <Dialog
-      contentClass="LMSFilePicker__dialog"
+    <Modal
+      contentClass={withFileUI ? 'LMSFilePicker' : ''}
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSFilePicker.js:341]: I've already got an issue tracking that `Modal` (and `Dialog`) use a legacy naming on this prop; it should be `classes`: https://github.com/hypothesis/frontend-shared/issues/181

File: lms/static/styles/_utils.scss [hunk 5]
```diff
@@ -19,3 +19,17 @@
...
+  text-overflow: ellipsis;
+  white-space: nowrap;
+}
+
+// A columnar flex layout centered on the cross axis
+.u-layout-column--align-center {
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+}
```
Comment (lyzadanger) [inline lms/static/styles/_utils.scss:35]: The idea here is to corral reusable utility-like classes for later evaluation for migration to the `frontend-shared` package (where they would be prefixed `.hyp-u-`). I've got some documentation about all of this started; hope to be able to share next week. Lots going on!

File: lms/static/styles/components/_Breadcrumbs.scss [hunk 6]
```diff
@@ -5,6 +5,9 @@
   // FIXME: For now, allow crumbs to wrap onto a second line if they
   // run out of room. This could be more elegant.
   flex-wrap: wrap;
+  // Avoid one-pixel height changes as breadcrumb items and dividers are
+  // added. FIXME: remove this rule once typography patterns have been established.
+  line-height: 1;
```
Comment (lyzadanger) [inline lms/static/styles/components/_Breadcrumbs.scss:10]: Don't even start me. But this fixes the problem.

File: lms/static/styles/components/_FileList.scss [hunk 7]
```diff
@@ -1,58 +1,16 @@
...
-  flex-direction: column;
-  vertical-align: middle;
-  text-align: center;
-  width: 100%;
-  min-height: 149px;
-  border: 1px solid var.$grey-6;
-  background-color: var.$grey-2;
+  // FIXME: Reduce specificity when possible to do so
+  // See https://github.com/hypothesis/frontend-shared/issues/180
+  & &__date-header {
```
Comment (lyzadanger) [inline lms/static/styles/components/_FileList.scss:6]: As linked, you see, I'll track this down and fix it in `frontend-shared` next week. Always learning...

File: lms/static/styles/components/_LMSFilePicker.scss [hunk 8]
```diff
@@ -0,0 +1,14 @@
+.LMSFilePicker {
+  // Set "preferred" height and width on the modal that displays a list of
+  // files in a table. (The Modal component styling also applies some max-width
+  // and max-height rules that will keep things within bounds). Setting these
+  // height and width rules keeps the file listing the same size no matter
+  // what its contents, and is larger than the default sizing of modals
+  // (because that is more convenient for browsing files).
+  height: min(90vh, 25rem);
+  width: min(90vw, 42rem);
+}
```
Comment (lyzadanger) [inline lms/static/styles/components/_LMSFilePicker.scss:10]: Let me know if this is confusing/unclear.

File: lms/static/scripts/frontend_apps/components/FileList.js [hunk 9]
```diff
@@ -41,49 +38,50 @@ export default function FileList({
...
+          accessibleLabel="File list"
+          classes="FileList"
+          tableHeaders={columns}
+          isLoading={isLoading}
+          items={files}
+          selectedItem={selectedFile}
+          onSelectItem={onSelectFile}
+          onUseItem={onUseFile}
+          renderItem={(file, isSelected) => (
+            <Fragment>
```
Comment (esanzgar) [inline lms/static/scripts/frontend_apps/components/FileList.js:63]:
```suggestion
            <>
```

File: lms/static/scripts/frontend_apps/components/FileList.js [hunk 10]
```diff
@@ -41,49 +38,50 @@ export default function FileList({
...
-            </td>
-          </Fragment>
-        )}
-      />
-      {!isLoading && files.length === 0 && noFilesMessage}
-      {isLoading && (
-        <div className="FileList__no-files-message">
-          <Spinner />
-        </div>
+    <Fragment>
```
Comment (esanzgar) [inline lms/static/scripts/frontend_apps/components/FileList.js:49]:
```suggestion
    <>
```

**Review by esanzgar — COMMENTED** [review]
I have no suggestion about the code, except some minor comments about inclusion of `<Fragment>`.

I am going to test this on LMS and then approve the PR.

File: lms/static/scripts/frontend_apps/components/FileList.js (hunk 9 above)
Comment (lyzadanger) [reply lms/static/scripts/frontend_apps/components/FileList.js:63]: Right. I wrote all of this code before you updated the `Fragment` configuration for this project...will have to fix throughout.

**Review by esanzgar — APPROVED** [review]
I tested in Blackboard and Canvas and all seems to work well.

**Comment by lyzadanger** [comment]
Thanks, @esanzgar ! I've just rebased to fix the conflicts. I'll likely deploy this tomorrow (8 Sep).

**Pull request merged**